node/out_response.js

```diff
@@ -256,25 +279,62 @@ TChannelOutResponse.prototype.send = function send(res1, res2) {
     self.arg3 = res2;
 
     if (self.ok) {
-        self.channel.inboundCallsSuccessStat.increment(1, {
-            'calling-service': self.inreq.headers.cn,
-            'service': self.inreq.serviceName,
-            'endpoint': String(self.inreq.arg1)
-        });
+        self.channel.emitFastStat(self.channel.buildStat(
+            'inbound.calls.success',
+            'counter',
+            1,
+            new InboundCallsSuccessTags(
+                self.inreq.headers.cn,
+                self.inreq.serviceName,
+                self.inreq.endpoint
+            )
+        ));
     } else {
         // TODO: add outResponse.setErrorType()
-        var type = 'unknown';
-        self.channel.inboundCallsAppErrorsStat.increment(1, {
-            'calling-service': self.inreq.headers.cn,
-            'service': self.inreq.serviceName,
-            'endpoint': String(self.inreq.arg1),
-            'type': type
-        });
+        self.channel.emitFastStat(self.channel.buildStat(
+            'inbound.calls.app-errors',
+            'counter',
+            1,
+            new InboundCallsAppErrorsTags(
+                self.inreq.headers.cn,
+                self.inreq.serviceName,
+                self.inreq.endpoint,
+                'unknown'
+            )
+        ));
     }
 
     self.sendCallResponseFrame([self.arg1, res1, res2], true);
     self.finishEvent.emit(self);
+
     return self;
 };
 
+function InboundCallsSuccessTags(cn, serviceName, endpoint) {
+    var self = this;
+
+    self.app = '';
+    self.host = '';
+    self.cluster = '';
+    self.version = '';
```

Comment:
These might benefit from a shared constructor for the common properties. Also, null might be faster than ''. I don’t believe V8 treats slots differently based on the type to which they are initially set.